batchglm/train/numpy/base_glm/modelContainer.py

```diff
@@ -0,0 +1,376 @@
+    ----------
+    theta_location : np.ndarray
+        Location model parameters
+    theta_scale : np.ndarray
+        Scale model parameters
+    converged : np.ndarray
+        Whether or not given parameters are converged
+    params : Union[np.ndarray, dask.array.core.Array]
+        Model parameters
+    converged : np.ndarray
+        Whether or not a parameter has converged
+    idx_train_loc : np.ndarray
+        Training indices for location model
+    idx_train_scale : np.ndarray
+        Training indices for scale model
+    npar_location : int
+        number of location parameters
+    dtype : str
+        data type to be used
+    """
+
+    params: Union[np.ndarray, dask.array.core.Array]
+    converged: np.ndarray
+    npar_location: int
+    dtype: str
+
+    def __init__(
+        self,
+        model,
+        init_theta_location: Union[np.ndarray, dask.array.core.Array],
+        init_theta_scale: Union[np.ndarray, dask.array.core.Array],
+        chunk_size_genes: int,
+        dtype: str,
+    ):
+        """
+        :param init_location:
+            Initialisation for all parameters of mean model. (mean model size x features)
+        :param init_scale:
+            Initialisation for all parameters of dispersion model. (dispersion model size x features)
+        :param chunk_size_genes:
+            chunk size for dask
+        :param dtype:
+            Precision used in tensorflow.
+        """
+
+        self.model = model
+        init_theta_location_clipped = model.np_clip_param(
+            np.asarray(init_theta_location, dtype=dtype), "theta_location"
+        )
+        init_theta_scale_clipped = model.np_clip_param(np.asarray(init_theta_scale, dtype=dtype), "theta_scale")
+        self.params = dask.array.from_array(
+            np.concatenate(
+                [
+                    init_theta_location_clipped,
+                    init_theta_scale_clipped,
+                ],
+                axis=0,
+            ),
+            chunks=(1000, chunk_size_genes),
+        )
+        self.npar_location = init_theta_location_clipped.shape[0]
+
+        # Properties to follow gene-wise convergence.
+        self.converged = np.repeat(a=False, repeats=self.params.shape[1])  # Initialise to non-converged.
+
+        self.dtype = dtype
+        self.idx_train_loc = np.arange(0, init_theta_location.shape[0])
+        self.idx_train_scale = np.arange(
+            init_theta_location.shape[0], init_theta_location.shape[0] + init_theta_scale.shape[0]
+        )
+
+        # overriding the location and scale parameter by referencing the getter functions within the properties.
+        self.model._theta_location_getter = self._theta_location_getter
+        self.model._theta_scale_getter = self._theta_scale_getter
+
+    def _theta_location_getter(self) -> dask.array.core.Array:
+        theta_location = self.params[0 : self.npar_location]
+        return self.np_clip_param(theta_location, "theta_location")
```

Comment:
No, the problem is that we need to update using the methods from the wrapped model object. But the model itself does not have access to the model_containers self.params. In the model, we already have the _theta_location_getter attribute. It is written in base_glm/model.py and by default points to _theta_location while here it is overriden to point to self.params. If we don't do this, self.params is used during the updates but the model's theta attributes are used for comparison between iterations.

Reply:
We could only circumvent this by multiple inheritance which I wanted to explicitly avoid which is why I used the wrapping approach here.

Reply:
The other way would be to have no redefinition of theta and instead use the ones in the model directly - sth that is cleaner but also a bit problematic with respect to separating model and fitting

Reply:
We also cannot directly point to the property because a reference to a property is not possible as it is simply creating a reference to the attribute by copying and returning it so it's not the same object in memory. This is why I added this wrapper method.

Reply:
Hmmmmm....the separation here is that on the model, the `theta_location` parameter is somehow static (i.e initial, and then updated after the final iteration) and here is where it is updated on every iteration? I guess I'm wondering why the model and this class both need `_theta_location_getter` especially since the one in `model.py` just returns the static parameter `_theta_location` by default.  Why does this behavior change and then is seemingly not changed back>

Reply:
If you call `estimator.model_container.model._theta_location` you get the original one.
If you call `estimator.model_container.theta_location` you get the params for location.

I agree on changing back the getter at the end. You could then call the initial and fitted using either
`estimator.model_container.theta_location` for fitted and
`estimator.model.theta_location` for the initial.

As I said, we need this getter because we cannot make a reference to a property.

Reply:
This is functional for now. And definitely an improvement over the previous multiple inheritance structure with references across parents. We leave it like that for now but come back to it once we start refacotring tf2 backend and adding poisson distribution.